File: sdk/samples/WFSamples/Applications/TrackingProfileDesigner/vb/WorkflowDesignerControl/WorkflowDesignerControl.vb

'---------------------------------------------------------------------
'   This file is part of the Windows Workflow Foundation SDK Code Samples.
' 
'   Copyright (C) Microsoft Corporation.  All rights reserved.
' 
' This source code is intended only as a supplement to Microsoft
' Development Tools and/or on-line documentation.  See these other
' materials for detailed information regarding Microsoft code samples.
' 
' THIS CODE AND INFORMATION ARE PROVIDED AS IS WITHOUT WARRANTY OF ANY
' KIND, EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE
' IMPLIED WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A
' PARTICULAR PURPOSE.
'---------------------------------------------------------------------

Imports System.ComponentModel.Design
Imports System.IO
Imports System.Workflow.Activities
Imports System.Workflow.ComponentModel
Imports System.Workflow.ComponentModel.Design
Imports System.Workflow.ComponentModel.Serialization
Imports System.XML

Public Class WorkflowDesignerControl
    Inherits UserControl
    Implements IDisposable, IServiceProvider

    Dim workflowView As WorkflowView
    Dim designSurface As DesignSurface
    Dim loader As WorkflowLoader

    Public Sub New()
        InitializeComponent()

        WorkflowTheme.CurrentTheme.ReadOnly = False
        WorkflowTheme.CurrentTheme.AmbientTheme.ShowConfigErrors = False
        WorkflowTheme.CurrentTheme.ReadOnly = True
    End Sub

    Dim xamlValue As String

    Public Property Xaml() As String
        Get
            Dim xamlScratch As String = String.Empty
            If Me.loader IsNot Nothing Then
                Me.loader.Flush()
                xamlScratch = Me.loader.Xaml
            End If
            Return xamlScratch
        End Get

        Set(ByVal value As String)
            Me.xamlValue = value
            If Not (String.IsNullOrEmpty(value)) Then
                LoadWorkflow()
            End If
        End Set
    End Property

    Dim workflowTypeValue As Type = Nothing

    Public Property WorkflowType() As Type
        Get
            Return workflowTypeValue
        End Get
        Set(ByVal value As Type)
            workflowTypeValue = value
            If value IsNot Nothing Then
                LoadWorkflow()
            End If
        End Set
    End Property

    Public Shadows Function GetService(ByVal serviceType As System.Type) As Object Implements System.IServiceProvider.GetService
        If Me.workflowView IsNot Nothing Then
            Return CType(Me.workflowView, IServiceProvider).GetService(serviceType)
        Else
            Return Nothing
        End If
    End Function

    Protected Overrides Sub OnLoad(ByVal e As EventArgs)
        MyBase.OnLoad(e)
    End Sub

    Private Sub LoadWorkflow()
        SuspendLayout()
        Dim designSurface As New DesignSurface()
        Dim loader As New WorkflowLoader()
        loader.WorkflowType = workflowTypeValue
        loader.Xaml = xamlValue
        designSurface.BeginLoad(loader)
        Dim designerHost As IDesignerHost = TryCast(designSurface.GetService(GetType(IDesignerHost)), IDesignerHost)
        If designerHost IsNot Nothing AndAlso designerHost.RootComponent IsNot Nothing Then
            Dim rootDesigner As IRootDesigner = TryCast(designerHost.GetDesigner(designerHost.RootComponent), IRootDesigner)
            If rootDesigner IsNot Nothing Then
                UnloadWorkflow()
                Me.designSurface = designSurface
                Me.loader = loader
                Me.workflowView = TryCast(rootDesigner.GetView(ViewTechnology.Default), WorkflowView)
                Me.panel1.Controls.Add(Me.workflowView)
                Me.workflowView.Dock = DockStyle.Fill
                Me.workflowView.TabIndex = 1
                Me.workflowView.TabStop = True
                Me.workflowView.HScrollBar.TabStop = False
                Me.workflowView.VScrollBar.TabStop = False
                Me.workflowView.Focus()
                Me.workflowView.FitToScreenSize()
            End If
        End If
        ResumeLayout(True)
    End Sub



    Custom Event SelectionChanged As EventHandler
        AddHandler(ByVal value As EventHandler)
            Dim selectionService As ISelectionService = TryCast(GetService(GetType(ISelectionService)), ISelectionService)
            If selectionService IsNot Nothing Then
                AddHandler selectionService.SelectionChanged, value
            End If
        End AddHandler

        RemoveHandler(ByVal value As EventHandler)
            Dim selectionService As ISelectionService = TryCast(GetService(GetType(ISelectionService)), ISelectionService)
            If selectionService IsNot Nothing Then
                RemoveHandler selectionService.SelectionChanged, value
            End If
        End RemoveHandler

        RaiseEvent(ByVal sender As Object, ByVal e As System.EventArgs)
        End RaiseEvent
    End Event

    Private Sub UnloadWorkflow()
        Dim designerHost As IDesignerHost = TryCast(GetService(GetType(IDesignerHost)), IDesignerHost)
        If designerHost IsNot Nothing AndAlso designerHost.Container.Components.Count > 0 Then
            WorkflowLoader.DestroyObjectGraphFromDesignerHost(designerHost, TryCast(designerHost.RootComponent, Activity))
        End If
        If Me.designSurface IsNot Nothing Then
            Me.designSurface.Dispose()
            Me.designSurface = Nothing
        End If

        If Me.workflowView IsNot Nothing Then
            Controls.Remove(Me.workflowView)
            Me.workflowView.Dispose()
            Me.workflowView = Nothing
        End If
    End Sub

    Public Sub ShowDefaultWorkflow()

        Dim workflow As New SequentialWorkflowActivity()
        workflow.Name = "Workflow1"

        Using stringWriter As New StringWriter()
            Using xmlWriter As XmlWriter = Xml.XmlWriter.Create(stringWriter)
                Dim serializer As New WorkflowMarkupSerializer()
                serializer.Serialize(xmlWriter, workflow)
                Me.Xaml = stringWriter.ToString()
            End Using
        End Using
    End Sub
End Class

File: sdk/samples/WFSamples/Applications/TrackingProfileDesigner/CS/WorkflowDesignerControl/WorkflowDesignerControl.cs
//---------------------------------------------------------------------
//  This file is part of the Windows Workflow Foundation SDK Code Samples.
// 
//  Copyright (C) Microsoft Corporation.  All rights reserved.
// 
//This source code is intended only as a supplement to Microsoft
//Development Tools and/or on-line documentation.  See these other
//materials for detailed information regarding Microsoft code samples.
// 
//THIS CODE AND INFORMATION ARE PROVIDED AS IS WITHOUT WARRANTY OF ANY
//KIND, EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE
//IMPLIED WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A
//PARTICULAR PURPOSE.
//---------------------------------------------------------------------


using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Drawing;
using System.Data;
using System.Text;
using System.Windows.Forms;
using System.ComponentModel.Design;
using System.ComponentModel.Design.Serialization;
using System.Workflow.ComponentModel;
using System.Workflow.ComponentModel.Compiler;
using System.Workflow.Activities;
using System.Workflow.ComponentModel.Design;
using System.Runtime.InteropServices;
using System.Collections;
using System.Collections.ObjectModel;
using System.IO;
using System.Xml;
using System.Workflow.ComponentModel.Serialization;

namespace Microsoft.Samples.Workflow.DesignerControl
{
    public partial class WorkflowDesignerControl : UserControl, IDisposable, IServiceProvider
    {
        private WorkflowView            workflowView;
        private DesignSurface           designSurface;
        private WorkflowLoader          loader;

        public WorkflowDesignerControl()
        {
            InitializeComponent();
            
            WorkflowTheme.CurrentTheme.ReadOnly = false;
            WorkflowTheme.CurrentTheme.AmbientTheme.ShowConfigErrors = false;
            WorkflowTheme.CurrentTheme.ReadOnly = true;            
        }

        private string xamlValue;

        public string Xaml
        {
            get
            {
                string xaml = string.Empty;
                if (this.loader != null)
                {
                    this.loader.Flush();
                    xaml = this.loader.Xaml;
                }
                return xaml;
            }

            set
            {
                this.xamlValue = value;
                if (!String.IsNullOrEmpty(value))
                    LoadWorkflow();
            }
        }

        private Type workflowTypeValue =null;

        public Type WorkflowType
        {
            get { return workflowTypeValue; }
            set { workflowTypeValue= value;
                if (value != null) LoadWorkflow();
            }
        }
    

        protected override void Dispose(bool disposing)
        {
            if (disposing && (components != null))
            {
                UnloadWorkflow();
                components.Dispose();
            }
            base.Dispose(disposing);
        }

        new public object GetService(Type serviceType)
        {
            return (this.workflowView != null) ? ((IServiceProvider)this.workflowView).GetService(serviceType) : null;
        }

        protected override void OnLoad(EventArgs e)
        {
            base.OnLoad(e);            
        }
        
        private void LoadWorkflow()
        {
            SuspendLayout();
            
            DesignSurface designSurface = new DesignSurface();
            WorkflowLoader loader = new WorkflowLoader();
            loader.WorkflowType = workflowTypeValue;
            loader.Xaml = xamlValue;            
            designSurface.BeginLoad(loader);

            IDesignerHost designerHost = designSurface.GetService(typeof(IDesignerHost)) as IDesignerHost;
            if (designerHost != null && designerHost.RootComponent != null)
            {
                IRootDesigner rootDesigner = designerHost.GetDesigner(designerHost.RootComponent) as IRootDesigner;
                if (rootDesigner != null)
                {
                    UnloadWorkflow();

                    this.designSurface = designSurface;
                    this.loader = loader;
                    this.workflowView = rootDesigner.GetView(ViewTechnology.Default) as WorkflowView;
                    this.panel1.Controls.Add(this.workflowView);
                    this.workflowView.Dock = DockStyle.Fill;
                    this.workflowView.TabIndex = 1;
                    this.workflowView.TabStop = true;
                    this.workflowView.HScrollBar.TabStop = false;
                    this.workflowView.VScrollBar.TabStop = false;
                    this.workflowView.Focus();
                    this.workflowView.FitToScreenSize();
                }
            }

            ResumeLayout(true);            
        }


        public event EventHandler SelectionChanged
        {
            add
            {
                ISelectionService selectionService = GetService(typeof(ISelectionService)) as ISelectionService;
                if (selectionService != null)
                    selectionService.SelectionChanged += new EventHandler(value);
            }
            remove
            {
                ISelectionService selectionService = GetService(typeof(ISelectionService)) as ISelectionService;
                if (selectionService != null)
                    selectionService.SelectionChanged -= new EventHandler(value);
            }
        }

        private void UnloadWorkflow()
        {
            IDesignerHost designerHost = GetService(typeof(IDesignerHost)) as IDesignerHost;
            if (designerHost != null && designerHost.Container.Components.Count > 0)
                WorkflowLoader.DestroyObjectGraphFromDesignerHost(designerHost, designerHost.RootComponent as Activity);
           
            if (this.designSurface != null)
            {
                this.designSurface.Dispose();
                this.designSurface = null;
            }

            if (this.workflowView != null)
            {
                Controls.Remove(this.workflowView);
                this.workflowView.Dispose();
                this.workflowView = null;
            }
        }

        public void ShowDefaultWorkflow()
        {
            SequentialWorkflowActivity workflow = new SequentialWorkflowActivity();
            workflow.Name = "Workflow1";

            using (StringWriter stringWriter = new StringWriter())
            {
                using (XmlWriter xmlWriter = XmlWriter.Create(stringWriter))
                {
                    WorkflowMarkupSerializer serializer = new WorkflowMarkupSerializer();
                    serializer.Serialize(xmlWriter, workflow);
                    this.Xaml = stringWriter.ToString();
                }
            }
        }
    }
}
